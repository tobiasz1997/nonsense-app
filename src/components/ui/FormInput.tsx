import cx from 'classnames';
import {
	forwardRef,
	ForwardRefRenderFunction,
	InputHTMLAttributes
} from 'react';
import FormErrorMessage from '@components/ui/FormErrorMessage';
import FormLabel from '@components/ui/FormLabel';

export type InputProps = {
	label?: string;
	error?: string;
	dimension?: sizeType;
};

const FormInput: ForwardRefRenderFunction<
	HTMLInputElement,
	InputProps & InputHTMLAttributes<HTMLInputElement>
> = (props, ref) => {
	const sizes: Record<sizeType, string> = {
		small: 'h-9',
		default: 'h-12'
	};

	const labelPosition: Record<sizeType, string> = {
		small: 'top-0 left-3',
		default: 'top-1 left-3'
	};

	return (
		<div>
			<label
				className={cx(
					'flex cursor-pointer items-center overflow-hidden rounded border focus-within:ring-4',
					props.error
						? 'border-red text-red focus-within:ring-red/10'
						: 'border-green-dark text-green-dark focus-within:ring-yellow'
				)}
			>
				<span className="relative w-full">
					{props.label && (
						<span
							className={cx(
								'absolute',
								labelPosition[props.dimension ?? 'default']
							)}
						>
							<FormLabel required={props.required}>{props.label}</FormLabel>
						</span>
					)}
					<input
						{...props}
						ref={ref}
						className={cx(
							'w-full border-none focus:outline-none focus:ring-0 disabled:opacity-50',
							props.label ? 'px-3 pb-2 pt-4' : 'p-2',
							sizes[props.dimension ?? 'default']
						)}
					/>
				</span>
			</label>
			{props.error && <FormErrorMessage>{props.error}</FormErrorMessage>}
		</div>
	);
};

type sizeType = 'default' | 'small';

export default forwardRef(FormInput);
