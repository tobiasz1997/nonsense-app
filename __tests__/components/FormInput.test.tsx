import { fireEvent, render, screen } from '@testing-library/react';
import FormInput from '@components/ui/FormInput';

describe('Form Input', () => {
	const content = '@#$%^&*()qwerty';

	it('should render without crashing', () => {
		const { container } = render(<FormInput />);
		expect(container).toMatchSnapshot();
	});

	it('should have correct value', async () => {
		render(<FormInput />);

		const input = screen.getByRole('textbox');
		fireEvent.change(input, { target: { value: content } });

		expect(input).toHaveDisplayValue(content);
	});
});
