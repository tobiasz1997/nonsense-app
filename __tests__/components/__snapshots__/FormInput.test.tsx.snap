// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Input should render without crashing 1`] = `
<div>
  <div>
    <label
      class="flex cursor-pointer items-center overflow-hidden rounded border focus-within:ring-4 border-green-dark text-green-dark focus-within:ring-yellow"
    >
      <span
        class="relative w-full"
      >
        <input
          class="h-12 w-full border-none focus:outline-none focus:ring-0 disabled:opacity-50 p-2"
        />
      </span>
    </label>
  </div>
</div>
`;
